common/src/main/java/com/microsoft/identity/common/internal/providers/microsoft/azureactivedirectory/AzureActiveDirectory.java

```diff
@@ -109,6 +126,39 @@ public static void initializeCloudMetadata(final String authorityHost, final Map
         }
     }
 
+    public static void performCloudDiscovery(){
+
+        Uri instanceDiscoveryRequestUri = Uri.parse(AAD_INSTANCE_DISCOVERY_ENDPOINT);
+
+        instanceDiscoveryRequestUri = instanceDiscoveryRequestUri
+                .buildUpon()
+                .appendQueryParameter(API_VERSION, API_VERSION_VALUE)
+                .appendQueryParameter(AUTHORIZATION_ENDPOINT, AUTHORIZATION_ENDPOINT_VALUE)
+                .build();
+
+        try {
+            HttpResponse response = HttpRequest.sendGet(new URL(instanceDiscoveryRequestUri.toString()),new HashMap<String, String>());
+
+            if(response.getStatusCode() >= HttpURLConnection.HTTP_BAD_REQUEST){
+                Log.d("Discovery", "Error getting cloud information");
+            }else{
+
+                AzureActiveDirectoryInstanceResponse instanceResponse =  ObjectMapper.deserializeJsonStringToObject(response.getBody(), AzureActiveDirectoryInstanceResponse.class);
+
+                for (final AzureActiveDirectoryCloud cloud : instanceResponse.getClouds()) {
+                    cloud.setIsValidated(true); // Mark the deserialized Clouds as validated
+                    for (final String alias : cloud.getHostAliases()) {
+                        sAadClouds.put(alias.toLowerCase(Locale.US), cloud);
+                    }
+                }
+            }
+        } catch (IOException e) {
```

Comment:
This failure case, if it happens, seems like a good candidate for telemetry as well